### PR TITLE
improve datapipe deprecation warnings

### LINKED ISSUE #322: Warnings for deprecated datapipes are insufficient
Consider this snippet:

```py
from torchdata.datapipes.iter import IterableWrapper

dp = IterableWrapper([])
dp = dp.read_from_tar()
```

There are three problems here:

1. Running this with `python $SCRIPT` will not give you any warning, because [`DeprecationWarning`](https://docs.python.org/3/library/exceptions.html#DeprecationWarning)'s are filtered by default. They are only visible if a user activates them manually, i.e. running `python -Wd $SCRIPT`:

    ```
    DeprecationWarning: _DataPipeMeta and its functional API are deprecated and will be removed from the package `torch`. Please use TarArchiveLoader instead.
    ```
    
    Since most users won't do that, we should using a [`FutureWarning`](https://docs.python.org/3/library/exceptions.html#FutureWarning), which is the user-facing deprecation warning.
2. Since the warning is emitted from `__new__`, we should use `cls.__name__`. Using `type(cls).__name__` will give us the name of the meta class as seen as `_DataPipeMeta` above.
3. If I used the functional API, it can be quite confusing what caused the warning. For example with the above fixed, the warning is

    ```
    FutureWarning: TarArchiveReaderIterDataPipe and its functional API are deprecated and will be removed from the package `torch`. Please use TarArchiveLoader instead.
    ```

    I have no `TarArchiveReaderIterDataPipe` in my code. Even if I had used the class interface, i.e. `from torchdata.datapipes.iter import TarArchiveReader`, there would still be a mismatch between my code and the warning.

I propose to switch `deprecation_warning` to something like

```py
def deprecation_warning(
    *,
    old_class_name: str,
    old_functional_name: str = "",
    new_class_name: str = "",
    new_functional_name: str = "",
) -> None:
    msg = f"`{old_class_name}()`"
    if old_functional_name:
        msg = f"{msg} and its functional API `.{old_functional_name}()`"
    # TODO: Make the deprecation and removal version concrete.
    #  See https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy#minimizing-the-disruption-of-bc-breaking-changes
    msg = f"{msg} are deprecated and will be removed in the future."

    if new_class_name or new_functional_name:
        msg = f"{msg} Please use"
        if new_class_name:
            msg = f"{msg} `{new_class_name}()`"
        if new_class_name and new_functional_name:
            msg = f"{msg} or"
        if new_functional_name:
            msg = f"{msg} `.{new_functional_name}()`"
        msg = f"{msg} instead."

    warnings.warn(msg, FutureWarning)
```

With this, the warning for the code above reads:

```
FutureWarning: `TarArchiveReaderIterDataPipe()` and its functional API `.read_from_tar()` are deprecated and will be removed in the future. Please use `TarArchiveLoader()` or `.load_from_tar()` instead.
```

LMK if I should send a PR to PyTorch core.

### PR DESCRIPTION
**Pull request by pmeier**
Fixes pytorch/data#322.

### CODE REVIEW

File: torch/utils/data/datapipes/utils/common.py
Comment (NivekT) [inline torch/utils/data/datapipes/utils/common.py:98]: Perhaps we can add the following link here?
```python
# https://github.com/pytorch/data/issues/163
```

**Review by NivekT — APPROVED** [review]
Thanks for doing this! LGTM with small comments

File: torch/utils/data/datapipes/utils/common.py
Comment (pmeier) [reply torch/utils/data/datapipes/utils/common.py]: Note that according to BC policy of PyTorch core, we can only remove functionality if the warning stating the concrete deprecation version was emitted for two releases. Thus, if we deprecate with the next releases `torch==1.12` / `torchdata==0.5`, we can only remove in `torch==1.14` / `torchdata==0.7`.

**Review by NivekT — APPROVED** [review]

**Comment by NivekT** [comment]
FYI, I am waiting for some PRs to land in `torchtext` to land (they will be switching away form deprecated APIs) before I can land this.

**Pull request closed without merging**
